onmt/modules/copy_generator.py

```diff
@@ -259,3 +237,69 @@ def _compute_loss(self, batch, output, target, copy_attn, align,
+                 lambda_coverage=0.0):
+        super(CopyGeneratorLossCompute, self).__init__(
+            criterion, generator, lambda_coverage=lambda_coverage)
+        self.tgt_vocab = tgt_vocab
+        self.normalize_by_length = normalize_by_length
+
+    def _make_shard_state(self, batch, output, range_, attns):
+        """See base class for args description."""
+        if getattr(batch, "alignment", None) is None:
+            raise AssertionError("using -copy_attn you need to pass in "
+                                 "-dynamic_dict during preprocess stage.")
+
+        shard_state = super(CopyGeneratorLossCompute, self)._make_shard_state(
+            batch, output, range_, attns)
+
+        shard_state.update({
+            "copy_attn": attns.get("copy"),
+            "align": batch.alignment[range_[0] + 1: range_[1]]
+        })
+        return shard_state
+
+    def _compute_loss(self, batch, output, target, copy_attn, align,
+                      std_attn=None, coverage_attn=None):
+        return CopyGeneratorLossComputeBase._compute_loss(self, batch, output,
+                                                          target, copy_attn,
+                                                          align, std_attn,
+                                                          coverage_attn)
+
+
+class CopyGeneratorLanguageModelLossCompute(LanguageModelLossCompute,
+                                            CopyGeneratorLossComputeBase):
+    """Copy Generator Loss Computation."""
+    def __init__(self, criterion, generator, tgt_vocab, normalize_by_length,
+                 lambda_coverage=0.0):
+        super(CopyGeneratorLanguageModelLossCompute, self).__init__(
+            criterion, generator, lambda_coverage=lambda_coverage)
+        self.tgt_vocab = tgt_vocab
+        self.normalize_by_length = normalize_by_length
+
+    def _compute_loss(self, batch, output, target, copy_attn, align,
+                      std_attn=None, coverage_attn=None):
+        return CopyGeneratorLossComputeBase._compute_loss(self, batch, output,
+                                                          target, copy_attn,
+                                                          align, std_attn,
+                                                          coverage_attn)
+
+    def _make_shard_state(self, batch, output, range_, attns):
+        """See base class for args description."""
+        if getattr(batch, "alignment", None) is None:
+            raise AssertionError("using -copy_attn you need to pass in "
+                                 "-dynamic_dict during preprocess stage.")
+
+        shard_state = super(CopyGeneratorLanguageModelLossCompute,
+                            self)._make_shard_state(batch, output, range_,
+                                                    attns)
+
+        shard_state.update({
+            "copy_attn": attns.get("copy"),
+            "align": batch.alignment[range_[0]: range_[1]]
```

Comment:
I agree that it is not a very nice approach. As well as the index shifting the two classes inherit two different losses. The difference in these losses are index shifting as well and the alignment loss.
It might be possible to keep one single class and to propagate the index shifting in a class attribute

`class CommonLossCompute(LossComputeBase):
    """
    Loss Computation parent for NMTLossCompute and LanguageModelLossCompute

    Implement loss compatible with coverage and alignement shards
    """

    def __init__(self, criterion, generator, normalization="sents",
                 lambda_coverage=0.0, lambda_align=0.0, tgt_index_shifting=1):
        super(CommonLossCompute, self).__init__(criterion, generator)
        self.lambda_coverage = lambda_coverage
        self.lambda_align = lambda_align
        self.tgt_index_shifting= tgt_index_shifting

class NMTLossCompute(CommonLossCompute):
    def __init__(self, criterion, generator, normalization="sents",
                 lambda_coverage=0.0, lambda_align=0.0):
        super(NMTLossCompute, self).__init__(criterion, generator, normalization="sents",
                 lambda_coverage=lambda_coverage, lambda_align=lambda_align, tgt_index_shifting=1)

class LanguageModelLossCompute(CommonLossCompute):
    def __init__(self, criterion, generator, normalization="sents",
                 lambda_coverage=0.0, lambda_align=0.0):
        super(LanguageModelLossCompute, self).__init__(criterion, generator, normalization="sents",
                 lambda_coverage=lambda_coverage, lambda_align=0.0, tgt_index_shifting=0)
`

and similar approach in CopyGeneratorLossCompute.

Do you prefer this solution ?